AVRO-3951: Document IDE support (#3366)
apache/avro: doc/content/en/docs/++version++/Editor Support/_index.md

---
title: "Editor Support"
linkTitle: "Editor Support"
weight: 209
---

<!--

 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   https://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.

-->

## Overview

> When editing an Avro schema, everyone has a preferred editor. Some use an IDE, like IntelliJ, Pycharm, RustRover or

Reviewer: ```suggestion
When editing an Avro schema, everyone has a preferred editor. Some use an IDE, like IntelliJ, PyCharm, RustRover or
```

> Visual Studio, while others prefer a more humble text editor like Atom/Pulsar, Emacs or Vim.

Reviewer: ```suggestion
Visual Studio, while others prefer a more humble text editor like Pulsar, Emacs or Vim.
```
IMO there is no need to mention Atom anymore since it is discontinued.

Author: Oops... I'm not used to the app... iI'll fix this tomorrow

## Emacs

Sometimes ridiculed as an operating system, Emacs is a rich text editor. Many packages are available
via one of the three main package archives: [GNU ELPA](https://elpa.gnu.org/), [non-GNU ELPA](https://elpa.nongnu.org/)
or [MELPA](https://melpa.org/).

### emacs-avro

[Package source](https://github.com/logc/emacs-avro)

Originally released elsewhere in 2013, the last update was in March 2021.

This package is not available via the main package archives.

Provides syntax highlighting for IDL schemata.

## JetBrains IDE family

JetBrains features an entire line of IDEs. They include IntelliJ IDEA, PyCharm, PhpStorm, GoLand, Rider, CLion,
RustRover, WebStorm, RubyMine,
DataGrip, DataSpell, ReSharper, Fleet, and Aqua.

All of them use the [Plugin Marketplace](https://plugins.jetbrains.com/) to load extensions from.

### Apache Avro IDL Schema Support

<img style="float: right; object-fit: contain; width: 160px; height: 160px;" src="jetbrainsAvroIdlSchemaSupport.svg"/>[Plugin](https://plugins.jetbrains.com/plugin/15728-apache-avro-idl-schema-support) ([source](https://github.com/opwvhk/avro-schema-support?tab=readme-ov-file#intellij-plugin-for-apache-avro-idl))

> Released in 2021, this plugin received its last update in November 2023.

Reviewer: ```suggestion
Released in 2021, this plugin received its last update in April 2025.
```

Features for Avro schema & protocol definitions:

* Syntax Highlighting
* Code Completion
* Code Formatting
* Error Highlighting
* Inspections & quick fixes (IDL only)
* New file templates

The plugin supports all JetBrains products.

### Avro and Parquet Viewer

<img style="float: right; object-fit: contain; width: 160px; height: 160px;" src="jetbrainsAvroAndParquetViewer.svg"/>[Plugin](https://plugins.jetbrains.com/plugin/12281-avro-and-parquet-viewer) ([source](https://github.com/benwatson528/intellij-avro-parquet-plugin))]

Released in 2021, this plugin received its last update in November 2022.

Allows previewing `.avro` files, and shows its schema in an editor tab.

### Big Data File Viewer

<img style="float: right; object-fit: contain; width: 160px; height: 160px;" src="jetbrainsBigDataFileViewer.svg"/>[Plugin](https://plugins.jetbrains.com/plugin/21701-big-data-file-viewer) (
part of the [Big Data Tools](https://plugins.jetbrains.com/bundles/8-big-data-tools))

Released in 2023, this plugin is actively developed by JetBrains.

Allows previewing `.avro` files, and shows its schema in the structure tool window.

## Pulsar/Atom

> Though the Atom.io editor has been sunset late 2022, the public fork [Pulsar](https://pulsar-edit.dev/) has taken its

Reviewer: 1. My earlier comment about Atom could be ignored!
2. Or we could drop it here too and mention just Pulsar.

Author: Thank you!
I'll go for option 2; that gives users more options.
